staging/src/k8s.io/pod-security-admission/admission/admission.go

```diff
@@ -0,0 +1,537 @@
+
+	podMetadata, podSpec, err := a.PodSpecExtractor.ExtractPodSpec(attrs.GetObject())
+	if err != nil {
+		klog.ErrorS(err, "failed to extract pod spec")
+		return internalErrorResponse("failed to extract pod template")
+	}
+	if podMetadata == nil && podSpec == nil {
+		// if a controller with an optional pod spec does not contain a pod spec, skip validation
+		return allowedResponse()
+	}
+	return a.CheckPod(ctx, podMetadata, podSpec, false)
+}
+
+// CheckPod looks up the policy for the pods namespace, and checks it against the given pod(-like) object.
+// The enforce policy is only checked if enforce=true.
+func (a *Admission) CheckPod(ctx context.Context, podMetadata *metav1.ObjectMeta, podSpec *corev1.PodSpec, enforce bool) admissionv1.AdmissionResponse {
+	// short-circuit on exempt runtimeclass
+	if a.exemptRuntimeClass(podSpec.RuntimeClassName) {
+		return allowedResponse()
+	}
+
+	namespace, err := a.NamespaceGetter.GetNamespace(ctx, podMetadata.Namespace)
+	if err != nil {
+		klog.ErrorS(err, "failed to fetch pod namespace", "namespace", podMetadata.Namespace)
+		return internalErrorResponse(fmt.Sprintf("failed to lookup namespace %s", podMetadata.Namespace))
+	}
+
+	auditAnnotations := map[string]string{}
+	policy, err := a.PolicyToEvaluate(namespace.Labels)
+	if err != nil {
+		klog.V(2).InfoS("failed to parse PodSecurity namespace labels", "err", err)
+		auditAnnotations["error"] = fmt.Sprintf("Failed to parse policy: %v", err)
+	}
+
+	response := allowedResponse()
+	if enforce {
+		if result, err := a.evaluateLevelAndVersion(podMetadata, podSpec, policy.Enforce); err != nil {
+			klog.ErrorS(err, "Enforce evaluation failed", "namespace", namespace.Name, "pod", podMetadata.Name, "profile", policy.Enforce.String())
+			response = internalErrorResponse(err.Error())
+		} else if !result.Allowed {
+			response = forbiddenResponse(result.ForbiddenDetail())
+		}
+	}
+
+	// TODO: reuse previous evaluation if audit level+version is the same as enforce level+version
+	if result, err := a.evaluateLevelAndVersion(podMetadata, podSpec, policy.Audit); err != nil {
+		klog.ErrorS(err, "Audit evaluation failed", "namespace", namespace.Name, "pod", podMetadata.Name, "profile", policy.Audit.String())
+		auditAnnotations["audit-error"] = err.Error()
+	} else if !result.Allowed {
+		auditAnnotations["audit"] = result.ForbiddenDetail()
+	}
+
+	// TODO: reuse previous evaluation if warn level+version is the same as audit or enforce level+version
+	if result, err := a.evaluateLevelAndVersion(podMetadata, podSpec, policy.Warn); err != nil {
+		klog.ErrorS(err, "Warn evaluation failed", "namespace", namespace.Name, "pod", podMetadata.Name, "profile", policy.Warn.String())
+	} else if !result.Allowed {
+		// TODO: Craft a better user-facing warning message
+		response.Warnings = append(response.Warnings, fmt.Sprintf("Pod violates PodSecurity profile %s: %s", policy.Warn.String(), result.ForbiddenDetail()))
+	}
+
+	response.AuditAnnotations = auditAnnotations
```

Comment:
nit: move up to the block where you set these?

Reply:
we set annotations in a couple different blocks (an error annotation if the namespace labels didn't parse right)... doing it at the end seems easier to reason about to me